[IV-23-23] Objetivo9

Morad02 opened the pull request:
API REST implementada mediante el uso de un  handler que se le pasa a Deno.Serve. Este handler recibe peticiones y en base al método usado en la petición y los parámetros ofrecidos usa una función o otra, como queda definido en routes. Para devolver la respuesta, se ha implementado una función llamada respond. En respond, determinamos los parámetros del http response, es decir, status, data..etc  Para el almacenamiento y gestión de los recursos se ha creado una especie de base de datos con Deno.Kv. Esta base de datos proporciona funciones para gestionar los recursos.

Morad02 committed:
[IV-23-23] Objetivo9

Morad02 committed:
Se usará nanoid closes #78

Se ha elegido esta libreria porque es frendly URL

Morad02 committed:
Funciones BD impelementadas closes #77

Se ha usado Deno.KV y nanoid

Morad02 committed:
BD testing closes #80

Testeando todas las funcionalidades (CRUD)

Morad02 committed:
funcionalidades API implementadas closes #81

Haciendo uso de las funciones de base de datos

Morad02 committed:
Impelentada HTTP response closes #82

Haciendo uso del resultado devuelto por las funciones de api y siguiendo buenas prácticas para status response

Morad02 committed:
API testing closes #84

Usando Deno.Serve y fetch

Morad02 committed:
Dockerfile modificado closes #83 #72

añadiendo: --unstable, --allow-net  test/config_ma_test.ts, test/db_test.ts, test/api_test.ts

Morad02 committed:
Añadidos install y build closes #85

Añadidos en el fichero deno.json

Morad02 committed:
Script para hacer el build closes #86

Usando las recomendaciones de Deno. Haciendo uso de esbuild

Morad02 committed:
Modificado import closes #86

haciendo uso del sistema de import, deno.json

Morad02 committed:
Añadido build y install closes #85

Haciendo uso de builds.ts y deno install

Morad02 committed:
Actualizads build y install cloese #85

En nuestro caso no es necesario

JJ commented on api/api.ts:
```diff
@@ -1,101 +1,133 @@
+            GET: async(id:string) => await db.default.getAsignacion(id),
             POST: async(id:string,params:any) => await postAsignacion(id,params),
-            DELETE: async(id:string) => await deleteAsignacion(id)
+            DELETE: async(id:string) => await db.default.deleteAsignacion(id)
```
¿La historia de usuario requiere que se borren todos los URIs?

Morad02 replied:
Resuelto en 0af3add8c195a9dda353c675990e3c159256b791

JJ commented on api/api.ts:
```diff
@@ -1,101 +1,133 @@
-}
+        if (method === 'GET' || method === 'DELETE' || method == 'UPDATE')
+            status = 404;
+        else if (method === 'CREATE' || method === 'ANY')
```
Estás mezclando verbos HTTP con otros valores que tú te has inventado. Eso es bastante peligroso. Por otro lado, acuérdate del single responsibility principle, que aquí se puede usar por varios lados: o se responsabiliza de los verbos HTTP, o se responsabiliza de otra cosa, pero también estás tratando de forma uniforme `data == null` y  `data == false`, que son cosas muy diferentes.

Morad02 replied:
Resuelto en 2ff06b1c2865315dcc8751bcf99896e85e855e88

JJ commented on api/api.ts:
```diff
@@ -1,101 +1,133 @@
 }
 
-async function deleteAsignacion(id:string)
+async function postAsignacion(id:string,params:any)
```
TypeScript tiene un sistema de tipos muy rico para que uses `any` en cualquier ocasión. ¿No has considerado usar los tipos que sean más adecuados en cada momento?

Morad02 replied:
Resuelto en b5bc2df2c004ca2ec6a953ab146afb3e2d9e28d9

JJ commented on api/api.ts:
```diff
@@ -1,101 +1,133 @@
+async function postAsignacion(id:string,params:any)
 {
-    
+    return id !== '' ? await db.default.updateAsignacion(id, params) : await db.default.createAsignacion(params);
```
Por otro lado, esto es repetición del código de las tres funciones anteriores; en lo único que varía es en la función que llama, pero la función que llama también tendrá el mismo código seguramente...

Morad02 replied:
Resuelto en 957a9f5a66a17bfd3f5ddd7695e080f2d8ae8ee5

JJ commented on model/db.ts:
```diff
@@ -1,86 +1,194 @@
 
-export interface Pedido {
+export interface Pedido 
+{
```
¿Por qué has cambiado esta convención?
Conviene en todo caso que estas cosas las hagáis antes del PR, porque genera cambios en ficheros que hay que evaluar de todas formas.

Morad02 replied:
El cambio no ha sido por alguna causa específica sino más bien porque se usaban las dos convenciones y se ha querido unificar. Perfecto, para las próximas veces lo tendré en cuenta.

JJ commented on model/db.ts:
```diff
@@ -1,86 +1,194 @@
-export interface Lote {
+import {nanoid} from "nanoid"
+import { Kv } from "../test/db_test.ts";
```
¿Estás importando del directorio de test en código de producción?

Morad02 replied:
resuelto en dd61fff6614c0d347323b1d391504d054a0c6a2b

JJ commented on test/db_test.ts:
```diff
@@ -0,0 +1,67 @@
+import * as db from '../model/db.ts';
+
+
+export const Kv = await Deno.openKv();
```
Ah, ya veo por qué. Esto no hace falta que lo exportes. Si te das cuenta, no tiene ningún tipo de parámetro, así que simplemente instanciarlo en cada módulo que necesites te devolverá la misma instancia. Mira el `saraer` mío y verás como lo hago.
Por otro lado, también conviene que uses la convención habitual en nombres en TS. Esto no es una clase, ¿no?

Morad02 replied:
Resuelto en dd61fff6614c0d347323b1d391504d054a0c6a2b

JJ reviewed (changes requested):
El objetivo está más o menos superado, pero hay una serie de buenas prácticas que son horizontales en todos los objetivos que me gustaría que siguieras.

Morad02 committed:
Quitado el borrado de recursos closes #87

Las historias de usuario no requiren el borrado de las URIs

Morad02 committed:
Creada getStatus closes #88

Creado esta función hacemos que responde tenga una única responsabilidad

Morad02 committed:
any remplazado closes #89

Sustituido por el tipado al que corresponde

Morad02 committed:
DRY post recurso api closes #90

Borradas las funciones post. EN su lugar, se llama a la función adecuada

Morad02 committed:
Eliminado Kv de los tests closes #91

Se usa solo en db, donde es necesario. Si se intancia más veces devuelve la primera creada

Morad02 commented:
@JJ cambios realizados en base a sus comentarios. Listo para revisión!

JJ commented on model/db.ts:
```diff
@@ -1,86 +1,159 @@
+    if (res.ok) {
+        return { item};
+    } else {
+        return null;
```
¿No has considerado la posibilidad de tirar una excepción que se capture a otro nivel? Como en el caso de los constructores, una función que cree un objeto debe *siempre* devolver objetos válidos. En este caso, además, no hay ninguna forma de que falle más que el que falle `Kv`, y eso sería suficientemente grave como para levantar una excepción.

JJ commented on model/db.ts:
```diff
@@ -1,86 +1,159 @@
     id: string;
     pesoMax: number;
     volumeMax: number;
-    lotes: string[];
```
¿Por qué has eliminado esto?

JJ commented on test/api_test.ts:
```diff
@@ -0,0 +1,115 @@
+    beforeAll(async () => {
+        conn = Deno.serve({ port: 8080 , handler: handler});
+        
+        lote = await fetch('http://localhost:8080/lote', {
```
Para testear no tienes por qué levantar el servidor en ningún momento. Deberías poder testear las rutas directamente.

JJ commented on test/api_test.ts:
```diff
@@ -0,0 +1,115 @@
+            body: JSON.stringify({params:{peso:100,volume:100}})
+        }).then( res => res.json() );
+
+        vehiculo = await fetch('http://localhost:8080/vehiculo', {
```
Entre otras cosas, ni siquiera sabes qué puertos hay disponibles. Y estás repitiendo una y otra vez un literal.

JJ reviewed (changes requested):
Dándole una vuelta, el problema principal es que no veo la conexión del API con la lógica de negocio. Te limitas a crear objetos y a recuperar objetos. El API debe servir al usuario exactamente siguiendo la HU de la misma forma, así que sí habría un pequeño obstáculo a la superación del objetivo, me temo...

JJ commented:
¿Qué queda por aquí?

Morad02 commented:
> ¿Qué queda por aquí?

Buenos días, la verdad es que he estado liado y no me ha sido posible volver a ponerme. Como me dijo, no estoy haciendo uso de la lógica de negocio. Para corregirlo supongo que tengo que pasar le al handler una instancia del gestorVehiculos para hacer uso de la lógica de negocio a la hora de asignar los pedidos a vehículos. Si encuentro un hueco me pondré con ello. Gracias!
